### PR TITLE
CVS-75092 Arbitrary batch size position

### PR DESCRIPTION
**Pull request by dkalinowski**
This commit implements missing logic for handling batch size on different position than 0.
Currently it was possible to declare arbitrary layout, for example `CHWN` and expect OVMS to treat 4th position as batch size.

This commit implements missing logic:
- handling request validation to report correct status - invalid shape or invalid batch size
- handling dynamic model parameters model reload when batch size is on position 1+
- handling DAG interconnections at runtime correctly - distinguish between invalid shape and invalid batch when batch size is on different position than 0

### CODE REVIEW

File: src/dlnodesession.cpp [hunk 1]
```diff
@@ -160,15 +160,23 @@ Status DLNodeSession::validate(const std::shared_ptr<ov::runtime::Tensor>& tenso
 
     // If batch size differs, check if remaining dimensions are equal
     const auto& dims = tensor->get_shape();
-    if (!tensorInfo.getShape()[0].match(dims[0])) {
+    const auto batchIndex = tensorInfo.getLayout().getBatchIndex();
+    if (!batchIndex.has_value() || tensorInfo.getShape().size() < batchIndex.value() + 1 || dims.size() < batchIndex.value() + 1) {
```
Comment (atobiszei) [inline src/dlnodesession.cpp:164]: MInor and could be skipped - wouldn't this be clearer?
```suggestion
    if (!batchIndex.has_value() ||
         batchIndex.value() >= tensorInfo.getShape().size() ||
         batchIndex.value() >=dims.size()) {
```
And anyway why we need to check if batchIndex.value() is not wrong within TensorInfo? At this point I think this shall not be needed - it may suggest that we missed something earlier on?

File: src/layout.cpp [hunk 2]
```diff
@@ -15,84 +15,62 @@
...
-
-#include "stringutils.hpp"
-
 namespace ovms {
 
-LayoutConfiguration::LayoutConfiguration(const char* layout) :
-    LayoutConfiguration(std::string(layout)) {
+Layout::Layout() :
+    std::string() {
 }
```
Comment (atobiszei) [inline src/layout.cpp:22]: Wouldn't delcaring Layout() = default suffice in header?

File: src/layout.cpp [hunk 3]
```diff
@@ -15,84 +15,62 @@
...
+        if (c == '.') {
+            if (etcAppeared) {
+                return StatusCode::LAYOUT_WRONG_FORMAT;  // Cannot appear multiple times
+            }
+            dotCount++;
+            if (dotCount >= 3) {
+                etcAppeared = true;
+            }
+        }
+    }
```
Comment (atobiszei) [inline src/layout.cpp:71]: WHat about handling eg NHW../NH.? Shouldn't those be illegal?
The same about N.H.W

File: src/layout_configuration.cpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+#include <algorithm>
+#include <sstream>
+#include <vector>
+
+#include "stringutils.hpp"
+
+namespace ovms {
+
+LayoutConfiguration::LayoutConfiguration(const char* layout) :
+    LayoutConfiguration(std::string(layout)) {
```
Comment (atobiszei) [inline src/layout_configuration.cpp:27]:
```suggestion
    LayoutConfiguration(layout) {
```
Wouldn't it work?

File: src/layout_configuration.cpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    LayoutConfiguration(layout, layout) {
+}
+
+LayoutConfiguration::LayoutConfiguration(const std::string& tensorLayout, const std::string& modelLayout) :
+    tensor(tensorLayout),
+    model(modelLayout) {
+}
+
+bool LayoutConfiguration::isSet() const {
+    return !tensor.empty() || !model.empty();
```
Comment (atobiszei) [inline src/layout_configuration.cpp:40]: THis return false if both are not set, if any is set it returns true. How we could construct it in such a way? Or is it just a precaution?

File: src/layout_configuration.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+Status LayoutConfiguration::fromString(const std::string& configurationStr, LayoutConfiguration& configOut) {
+    std::string configurationCopy = configurationStr;
+    erase_spaces(configurationCopy);
+
+    std::transform(configurationCopy.begin(), configurationCopy.end(), configurationCopy.begin(), ::toupper);
+
+    if (configurationCopy.find_first_not_of("NCHWD?.:") != std::string::npos)
+        return StatusCode::LAYOUT_WRONG_FORMAT;
```
Comment (atobiszei) [inline src/layout_configuration.cpp:50]: Maybe we could define
const char LAYOUT_CONFIG_DELIMITER = ":";
const std::string ALLOWED_LAYOUT_CHARS = "NCHWD?.";
const  std::string ALLOWED_LAYOUT_CONFIG_CHARS = ALLOWED_LAYOUT_CHARS + LAYOUT_CONFIG_DELIMITER;
etc?

File: src/layout_configuration.hpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    bool isSet() const;
+
+    bool operator==(const LayoutConfiguration& rhs) const;
+    bool operator!=(const LayoutConfiguration& rhs) const;
+
+    static Status fromString(const std::string& configurationStr, LayoutConfiguration& configOut);
+    std::string toString() const;
+};
+
+using layouts_map_t = std::unordered_map<std::string, LayoutConfiguration>;
```
Comment (atobiszei) [inline src/layout_configuration.hpp:47]: layout_configuration_map_t?

File: src/test/predict_validation_test.cpp [hunk 8]
```diff
@@ -468,19 +469,94 @@ TEST_F(PredictValidation, RequestIncorrectValueCountShapeAuto) {
...
     auto status = instance->mockValidate(&request);
     EXPECT_EQ(status, ovms::StatusCode::INVALID_SHAPE);
 }
 
+class PredictValidationArbitraryBatchPosition : public PredictValidation {
+protected:
+    void SetUp() override {
+        PredictValidation::SetUp();
+
+        networkInputs = ovms::tensor_map_t({
```
Comment (atobiszei) [inline src/test/predict_validation_test.cpp:492]: Depending on which MR comes first - this possibly should be servableInputs

File: src/test/shape_test.cpp [hunk 9]
```diff
@@ -54,8 +54,8 @@ TEST(Shape, OvShapeMatch) {
     EXPECT_FALSE(ovms::Shape({1, 6, 8}).match(ov::Shape({1, 6})));
     EXPECT_TRUE(ovms::Shape({1, 6, 8}).match(ov::Shape({1, 6, 8})));
     EXPECT_TRUE(ovms::Shape({{1, 2}, {6, 12}, Dimension::any()}).match(ov::Shape({1, 8, 100})));
-    size_t startingPosition = 2;
-    EXPECT_TRUE(ovms::Shape({{3, 5}, {7, 10}, {11, 19}}).match(ov::Shape({1000, 1000, 12}), startingPosition));
-    startingPosition = 200;
-    EXPECT_TRUE(ovms::Shape({{3, 5}, {7, 10}, {11, 19}}).match(ov::Shape({1000, 1000, 12000}), startingPosition));
+    size_t skipPosition = 1;
```
Comment (atobiszei) [inline src/test/shape_test.cpp:57]: Why this change? We started comparison from startingPosition and continued onwards. Skip position suggests that we skip just this one position (which is not true - we start comparison just after this position).

File: src/dlnodesession.cpp (hunk 1 above)
Comment (dkalinowski) [reply src/dlnodesession.cpp:164]: TODO: Remove shape size vs dims size batch size index. To be done in separate task in next sprint (todo: update task description)

File: src/test/layout_test.cpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+TEST(Layout, BatchPositionValid) {
+    EXPECT_EQ(Layout{"NHWC"}.getBatchIndex(), 0);
+    EXPECT_EQ(Layout{"HWCN"}.getBatchIndex(), 3);
+    EXPECT_EQ(Layout{"NC"}.getBatchIndex(), 0);
+    EXPECT_EQ(Layout{"NCHW"}.getBatchIndex(), 0);
+    EXPECT_EQ(Layout{"CHW"}.getBatchIndex(), std::nullopt);
+    EXPECT_EQ(Layout{"N..."}.getBatchIndex(), 0);
+    EXPECT_EQ(Layout{"N...CH"}.getBatchIndex(), 0);
+    EXPECT_EQ(Layout{"??N..."}.getBatchIndex(), 2);
+    EXPECT_EQ(Layout{"?C???N..."}.getBatchIndex(), 5);
```
Comment (dtrawins) [inline src/test/layout_test.cpp:34]: what would happen with Layout{"...NC"}

File: src/layout_configuration.cpp (hunk 4 above)
Comment (dkalinowski) [reply src/layout_configuration.cpp:27]: That would end up being endless recursive call:
```
error: constructor delegates to itself
```

File: src/layout_configuration.cpp (hunk 5 above)
Comment (dkalinowski) [reply src/layout_configuration.cpp:40]: This is exactly what user specifies in CLI or config.json. The tensor and model field is empty, when user did not specify layout in CLI or config.json. This is then translated to default `N...` during model loading.

File: src/test/layout_test.cpp (hunk 10 above)
Comment (dkalinowski) [reply src/test/layout_test.cpp:34]: done

File: src/test/predict_validation_test.cpp (hunk 8 above)
Comment (dkalinowski) [reply src/test/predict_validation_test.cpp]: done

File: src/layout_configuration.hpp (hunk 7 above)
Comment (dkalinowski) [reply src/layout_configuration.hpp]: done

File: src/layout.cpp (hunk 3 above)
Comment (dkalinowski) [reply src/layout.cpp:72]: done

File: src/layout.cpp (hunk 2 above)
Comment (dkalinowski) [reply src/layout.cpp]: done

File: src/dlnodesession.cpp (hunk 1 above)
Comment (dkalinowski) [reply src/dlnodesession.cpp]: done

**Review by dtrawins — APPROVED** [review]

**Review by atobiszei — APPROVED** [review]
